### PR TITLE
add notification module for Microsoft Teams

### PR DESCRIPTION
**Pull request by diadzine**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
teams

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel 0224760725) last updated 2017/03/02 23:56:33 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY
New module to send notification to Microsoft Teams.

### CODE REVIEW

**Comment by mattclay** [comment]
The test `ansible-test sanity --test validate-modules` failed with the following errors:

```            
lib/ansible/modules/notification/teams.py:0:0: E312 No RETURN documentation provided
lib/ansible/modules/notification/teams.py:0:0: E208 module_utils imports should import specific components, not "*". line 139
lib/ansible/modules/notification/teams.py:0:0: E208 module_utils imports should import specific components, not "*". line 140
lib/ansible/modules/notification/teams.py:0:0: E107 Imports should be directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 139
lib/ansible/modules/notification/teams.py:0:0: E107 Imports should be directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 140
```

File: lib/ansible/modules/notification/teams.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        The simplest message just contains a text variable.
+        To add links to text, developers **must** include Markdown syntax for
+        hyperlinks.
+    required: true
+  color:
+    description:
+      - You can also control a color bar by setting color to a color hex code.
+        This bar appears to the left of the cards content, and is primarily
+        used to indicate status to the user.
+    required: false
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Parameters that are not required don't need `required: false`.

File: lib/ansible/modules/notification/teams.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    default: None
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: If there's no default, `None' is implicit. (It means parameter undefined)
So this is not needed.

File: lib/ansible/modules/notification/teams.py (hunk 1 above)
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Parameters that are not required don't need `required: false`.

File: lib/ansible/modules/notification/teams.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            )
+        )
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            url = dict(type='str', required=True, no_log=True),
+            text = dict(type='str', required=True),
+            title = dict(type='str', required=False, default=None),
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Same here. required=False and default=None are obsolete. Only specify them if they are set.

File: lib/ansible/modules/notification/teams.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            color = dict(type='str', required=False, default=None)
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Same here. required=False and default=None are obsolete. Only specify them if they are set.

File: lib/ansible/modules/notification/teams.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        )
+    )
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Please implement check-mode. In this case check-mode could either not send a notification, or better, sends a "fake" notification (so authentication is attempted). It depends on the API if this is feasible. It's always better that check-mode is performing more logic without actually doing something.

File: lib/ansible/modules/notification/teams.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    - The C(teams) module sends notifications to
+      U(https://teams.microsoft.com/start) through Office 365 Connectors via
+      the Incoming WebHook integration.
+version_added: "2.3"
+author: "Aymeric Bringard (@diadzine)"
+options:
+  title:
+    description:
+      - A message can also contain a title.
+    required: false
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: Parameters that are not required don't need `required: false`.

File: lib/ansible/modules/notification/teams.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    default: None
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: If there's no default, `None' is implicit. (It means parameter undefined)
So this is not needed.

File: lib/ansible/modules/notification/teams.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+    url = module.params['url']
+    text = module.params['text']
+    title = module.params['title']
+    color = module.params['color']
+
+    payload = build_payload_for_teams(module, text, title, color)
+    do_notify_teams(module, url, payload)
+
+    module.exit_json(msg="OK")
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py]: What's the purpose of sending `msg="OK"` if the success of the task indicates the same thing ?

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by bjolivot** [comment]
Thank you for your module :+1: 
shipit

File: lib/ansible/modules/notification/teams.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        headers=headers,
+        method='POST',
+        data=payload
+    )
+
+    if info['status'] == 200:
+        module.exit_json(changed=True)
+    else:
+        module.fail_json(
+            msg=" failed to send %s to Teams incoming webhook: %s" % (
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:116]: Not sure why the error starts with a whitespace here.

**Comment by gundalow** [comment]
Hi, there is an alternative version https://github.com/ansible/ansible/pull/23277

Would be good if between you both you could decide which should go forward.

**Comment by marc-sensenich** [comment]
@diadzine I am happy to share ownership and maintenance moving forward on this module's functionality and am open to merging whichever PR accomplishes this best.

**Comment by northalpha** [comment]
@marc-sensenich could you please include your message cards support? that would be very nice! thanks!

File: lib/ansible/modules/notification/teams.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+DOCUMENTATION = """
+module: teams
+short_description: Send Microsoft Teams notifications
+description:
+    - The C(teams) module sends notifications to
+      U(https://teams.microsoft.com/start) through Office 365 Connectors via
+      the Incoming WebHook integration.
+version_added: "2.3"
+author: "Aymeric Bringard (@diadzine)"
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:33]: Make this a proper list:
```yaml
author:
- Aymeric Bringard (@diadzine)
```

File: lib/ansible/modules/notification/teams.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        ),
+        supports_check_mode=True
+    )
+
+    url = module.params['url']
+    text = module.params['text']
+    title = module.params['title']
+    color = module.params['color']
+
+    if module.check_mode:
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:139]: Move the check_mode test into the `do_notify_teams` function, and see if you can do something relevant instead e.g. test the connectivity so you do as much as possible in check-mode.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
If you want to speed up getting this merged, put it on the agenda of the core meeting and ask for a review there.

+label new_module

**Comment by marc-sensenich** [comment]
@dagwieers this isn't my pull request, I ended up opening #23277 before I saw this one which implemented all the features of the 365 connector cards at the time. I'm willing to reopen a PR with my module implementation and put it on the agenda since this PR has gone stale. Where would I submit an agenda item?

**Comment by dagwieers** [comment]
@marc-sensenich Did you talk to @diadzine ? It's best to come to an agreement.

**Comment by marc-sensenich** [comment]
I reached out in April in this PR as a comment after I opened mine and had no response.

**Comment by dagwieers** [comment]
@marc-sensenich I feel your pain. So maybe open your PR again, and work on that one. Then add a comment on the agenda about this situation and we will discuss in the meeting when you join.

The meeting agenda is here: https://github.com/ansible/community/issues/162

**Comment by jhawkesworth** [comment]
Thanks for this module.  

I just tried this out and it works for me so it would be great to get this (or a more enhanced version) included in ansible.

**Comment by jhawkesworth** [comment]
@diadzine - Are you still interested in getting this module included in ansible?  There's a little bit of time to respond to the feedback above and fix this up before ansible 2.4 deadline.  If you don't have time for this please can you respond as there are others willing to make and maintain a module for Teams.  Many thanks, Jon

**Comment by s-hertel** [comment]
@diadzine @dagwieers @jhawkesworth 
Since there has been no activity updating this since March and it would be a good thing to make it into 2.4 and @marc-sensenich's PR in addition has more functionality, I would like to merge #27636 (after pep8 is fixed). Are there any objections, or do people still want a chance to review it?

File: lib/ansible/modules/notification/teams.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/python
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:1]: Remove this shebang (CI will fail)

File: lib/ansible/modules/notification/teams.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+# -*- coding: utf-8 -*-
+
+# (c) 2017, Aymeric Bringard <diadzine@gmail.com>
+#
+# This file is part of Ansible
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:6]: We now have a one-line GPL stub for modules. See other modules in devel branch.

File: lib/ansible/modules/notification/teams.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:21]: This is the old ANSIBLE_METADATA stub. CI will fail.

File: lib/ansible/modules/notification/teams.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+options:
+  title:
+    description:
+      - A message can also contain a title.
+  text:
+    description:
+      - The contents of text will be displayed in a card format, along with a
+        sender image and display name. This image and display name are set
+        when the webhook is configured.
+        The simplest message just contains a text variable.
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:43]: Make this a separate item in the list.

File: lib/ansible/modules/notification/teams.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        To add links to text, developers **must** include Markdown syntax for
```
Comment (dagwieers) [inline lib/ansible/modules/notification/teams.py:44]: Make this a separate item in the list.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
@s-hertel I had some more unanswered comments, and added a few more (related to future CI failures). But I cannot test the module itself. For me this is fine to be merged if it works for others.

**Comment by s-hertel** [comment]
@dagwieers in regard to #27636? I think #27636 should be merged since marc-sensenich is active and happy to make fixes.

**Comment by dagwieers** [comment]
@s-hertel Yes, sorry somehow I misread your comment. This one can be closed IMO.

**Comment by diadzine** [comment]
Sorry guys, I was busy those last months with no time on this PR.
As @s-hertel suggested #27636 has more functionality and @marc-sensenich is more active than me ;)
I close this PR.

**Pull request closed without merging**